base/iddict.jl

```diff
@@ -83,18 +84,31 @@ function sizehint!(d::IdDict, newsz)
     rehash!(d, newsz)
 end
 
-function setindex!(d::IdDict{K,V}, @nospecialize(val), @nospecialize(key)) where {K, V}
+function ht_keyindex!(d::IdDict{K,V}, @nospecialize(key)) where {K, V}
     !isa(key, K) && throw(KeyTypeError(K, key))
-    if !(val isa V) # avoid a dynamic call
-        val = convert(V, val)::V
-    end
+    keyindex = ccall(:jl_eqtable_keyindex, Cssize_t, (Any, Any), d, key)
+    # keyindex - where a key is stored, or -pos if the key was not present and was inserted at pos
+
+    return abs(keyindex), keyindex < 0
+end
+
+function _setindex!(d::IdDict{K,V}, val::V, keyindex::Int, inserted::Bool) where {K, V}
+    @inbounds d.ht[keyindex+1] = val
+    d.count += inserted
+
     if d.ndel >= ((3*length(d.ht))>>2)
         rehash!(d, max((length(d.ht)%UInt)>>1, 32))
         d.ndel = 0
     end
-    inserted = RefValue{Cint}(0)
-    d.ht = ccall(:jl_eqtable_put, Memory{Any}, (Any, Any, Any, Ptr{Cint}), d.ht, key, val, inserted)
-    d.count += inserted[]
+    return nothing
+end
+
+function setindex!(d::IdDict{K,V}, @nospecialize(val), @nospecialize(key)) where {K, V}
+    keyindex, inserted = ht_keyindex!(d, key)
+    if !(val isa V) # avoid a dynamic call
+        val = convert(V, val)::V
```

Comment:
There seems to be a discrepancy here in the state of the IdDict during this convert call or if this convert failed. The `ht_keyindex!` call has fully inserted the key, but not the value. For `Dict`, neither has occurred yet, and all insertion and accounting happens at once in `_setindex!`. But in IdDict here, it looks like half of the insertion occurs in `ht_keyindex!` (effectively inserting it in the deleted state) but then the final accounting update for it doesn't happen until after the `_setindex!` call later in this method. Do we update `ndel` over this call so the value is consistent?

There also may need to be an `age` counter added to the IdDict, so it can detect concurrent modifications that happen in this `convert` call (or move this convert call back to the top of the function)